bench-run/src/main/scala/dotty/tools/benchmarks/specialization/Functions.scala

```diff
@@ -0,0 +1,55 @@
+package dotty.tools.benchmarks.specialization
+
+import org.openjdk.jmh.annotations._
+import scala.util.Random
+
+class Functions {
+  extension (x: Int)
+    inline def times(op: => Unit): Unit = {
+      var count = 0
+      while count < x do
+        op
+        count += x
+    }
+
+  class ByName {
+    def foo(x: => Int): Int = x
+  }
+
+  @Benchmark
+  def byNameBench() = 10000.times {
+    val a = new ByName
+    var list = List(a)
+    list.head.foo(6) + 10
+  }
+
+  @Benchmark
+  def lambdaBench() = 10000.times {
+    val fn = (x: Int) => x + 1
+    var list = List(fn)
+    list.head(2)
+  }
+
+  class Func1[T](fn: T => Int) extends Function1[T, Int] {
+    def apply(x: T): Int = fn(x)
+  }
+  class Fn extends Func1(identity[Int])
+
+  @Benchmark
+  def extendFun1Bench() = 10000.times {
+    val a: Function1[Int, Int] = new Fn
+    var list = List(a)
+    list.head(123) + 10
```

Comment:
As far as I can tell, specialization will only affect the last line in the `times` operation. Maybe we compute that `list` outside the `times` to have more precise measurements of the operations. In this scenario, if specialization works, not boxing or other allocation should happen in the loop which will be simpler to distinguish from one that does allocate.

Reply:
I'll try that. The only worry is that JVM is so smart that it aggressively inlines everything thus no boxing happens at all.

Reply:
Right. Actually the current code could also suffer from this. There is a trick to avoid that by making sure the result of the computation is needed.
```scala
@Benchmark
  def extendFun1Bench(): Int =
    val a: Function1[Int, Int] = new Fn
    var list = List(a)
    var res = 0
    10000.times {
      res += list.head(123) + 10
    }
    res
}
```

Reply:
The Tuple benchmarks always return the result for this purpose.

Reply:
That is useful, thanks for the tip 👍